// @iota/iota-interaction-ts/node/move_calls/identity/borrow_asset.js
"use strict";
// Copyright 2020-2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.proposeBorrow = proposeBorrow;
exports.executeBorrow = executeBorrow;
exports.createAndExecuteBorrow = createAndExecuteBorrow;
const transactions_1 = require("@iota/iota-sdk/transactions");
const utils_1 = require("../utils");
function proposeBorrow(identity, capability, objects, packageId, expiration) {
    const tx = new transactions_1.Transaction();
    const cap = tx.objectRef(capability);
    const [delegationToken, borrow] = (0, utils_1.getControllerDelegation)(tx, cap, packageId);
    const identityArg = tx.sharedObjectRef(identity);
    const exp = tx.pure.option("u64", expiration);
    const objectsArg = tx.pure.vector("id", objects);
    tx.moveCall({
        target: `${packageId}::identity::propose_borrow`,
        arguments: [identityArg, delegationToken, exp, objectsArg],
    });
    (0, utils_1.putBackDelegationToken)(tx, cap, delegationToken, borrow, packageId);
    return tx.build();
}
function executeBorrow(identity, capability, proposalId, objects, intentFn, packageId) {
    const tx = new transactions_1.Transaction();
    const cap = tx.objectRef(capability);
    const [delegationToken, borrow] = (0, utils_1.getControllerDelegation)(tx, cap, packageId);
    const proposal = tx.pure.id(proposalId);
    const identityArg = tx.sharedObjectRef(identity);
    let action = tx.moveCall({
        target: `${packageId}::identity::execute_proposal`,
        typeArguments: [`${packageId}::borrow_proposal::Borrow`],
        arguments: [identityArg, delegationToken, proposal],
    });
    (0, utils_1.putBackDelegationToken)(tx, cap, delegationToken, borrow, packageId);
    const objectArgMap = new Map();
    for (const obj of objects) {
        const recvObj = tx.receivingRef(obj);
        const objArg = tx.moveCall({
            target: `${packageId}::identity::execute_borrow`,
            typeArguments: [obj.type],
            arguments: [identityArg, action, recvObj],
        });
        objectArgMap.set(obj.objectId, [objArg, obj]);
    }
    intentFn(tx, objectArgMap);
    for (const [obj, objData] of objectArgMap.values()) {
        tx.moveCall({
            target: `${packageId}::borrow_proposal::put_back`,
            typeArguments: [objData.type],
            arguments: [action, obj],
        });
    }
    tx.moveCall({
        target: `${packageId}::transfer_proposal::conclude_borrow`,
        arguments: [action],
    });
    return tx.build();
}
function createAndExecuteBorrow(identity, capability, objects, intentFn, packageId, expiration) {
    const tx = new transactions_1.Transaction();
    const cap = tx.objectRef(capability);
    const [delegationToken, borrow] = (0, utils_1.getControllerDelegation)(tx, cap, packageId);
    const identityArg = tx.sharedObjectRef(identity);
    const exp = tx.pure.option("u64", expiration);
    const objectsArg = tx.pure.vector("id", objects.map(obj => obj.objectId));
    const proposal = tx.moveCall({
        target: `${packageId}::identity::propose_borrow`,
        arguments: [identityArg, delegationToken, exp, objectsArg],
    });
    let action = tx.moveCall({
        target: `${packageId}::identity::execute_proposal`,
        typeArguments: [`${packageId}::borrow_proposal::Borrow`],
        arguments: [identityArg, delegationToken, proposal],
    });
    (0, utils_1.putBackDelegationToken)(tx, cap, delegationToken, borrow, packageId);
    const objectArgMap = new Map();
    for (const obj of objects) {
        const recvObj = tx.receivingRef(obj);
        const objArg = tx.moveCall({
            target: `${packageId}::identity::execute_borrow`,
            typeArguments: [obj.type],
            arguments: [identityArg, action, recvObj],
        });
        objectArgMap.set(obj.objectId, [objArg, obj]);
    }
    intentFn(tx, objectArgMap);
    for (const [obj, objData] of objectArgMap.values()) {
        tx.moveCall({
            target: `${packageId}::borrow_proposal::put_back`,
            typeArguments: [objData.type],
            arguments: [action, obj],
        });
    }
    tx.moveCall({
        target: `${packageId}::transfer_proposal::conclude_borrow`,
        arguments: [action],
    });
    return tx.build();
}
