// @iota/iota-interaction-ts/node/move_calls/identity/controller_execution.js
"use strict";
// Copyright 2020-2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.proposeControllerExecution = proposeControllerExecution;
exports.executeControllerExecution = executeControllerExecution;
exports.createAndExecuteControllerExecution = createAndExecuteControllerExecution;
const transactions_1 = require("@iota/iota-sdk/transactions");
const utils_1 = require("../utils");
function proposeControllerExecution(identity, capability, controllerCapId, packageId, expiration) {
    const tx = new transactions_1.Transaction();
    const cap = tx.objectRef(capability);
    const [delegationToken, borrow] = (0, utils_1.getControllerDelegation)(tx, cap, packageId);
    const identityArg = tx.sharedObjectRef(identity);
    const exp = tx.pure.option("u64", expiration);
    const controllerCapIdArg = tx.pure.id(controllerCapId);
    tx.moveCall({
        target: `${packageId}::identity::propose_controller_execution`,
        arguments: [identityArg, delegationToken, controllerCapIdArg, exp],
    });
    (0, utils_1.putBackDelegationToken)(tx, cap, delegationToken, borrow, packageId);
    return tx.build();
}
function executeControllerExecution(identity, capability, proposalId, controllerCapRef, intentFn, packageId) {
    const tx = new transactions_1.Transaction();
    const cap = tx.objectRef(capability);
    const [delegationToken, borrow] = (0, utils_1.getControllerDelegation)(tx, cap, packageId);
    const proposal = tx.pure.id(proposalId);
    const identityArg = tx.sharedObjectRef(identity);
    let action = tx.moveCall({
        target: `${packageId}::identity::execute_proposal`,
        typeArguments: [`${packageId}::borrow_proposal::Borrow`],
        arguments: [identityArg, delegationToken, proposal],
    });
    (0, utils_1.putBackDelegationToken)(tx, cap, delegationToken, borrow, packageId);
    const receiving = tx.receivingRef(controllerCapRef);
    const BorrowedControllerCap = tx.moveCall({
        target: `${packageId}::identity::borrow_controller_cap`,
        arguments: [identityArg, action, receiving],
    });
    intentFn(tx, BorrowedControllerCap);
    tx.moveCall({
        target: `${packageId}::controller_proposal::put_back`,
        arguments: [action, BorrowedControllerCap],
    });
    return tx.build();
}
function createAndExecuteControllerExecution(identity, capability, controllerCapRef, intentFn, packageId, expiration) {
    const tx = new transactions_1.Transaction();
    const cap = tx.objectRef(capability);
    const [delegationToken, borrow] = (0, utils_1.getControllerDelegation)(tx, cap, packageId);
    const identityArg = tx.sharedObjectRef(identity);
    const exp = tx.pure.option("u64", expiration);
    const controller_cap_id = tx.pure.id(controllerCapRef.objectId);
    const proposal = tx.moveCall({
        target: `${packageId}::identity::propose_controller_execution`,
        arguments: [identityArg, delegationToken, controller_cap_id, exp],
    });
    let action = tx.moveCall({
        target: `${packageId}::identity::execute_proposal`,
        typeArguments: [`${packageId}::borrow_proposal::Borrow`],
        arguments: [identityArg, delegationToken, proposal],
    });
    (0, utils_1.putBackDelegationToken)(tx, cap, delegationToken, borrow, packageId);
    const receiving = tx.receivingRef(controllerCapRef);
    const borrowedControllerCap = tx.moveCall({
        target: `${packageId}::identity::borrow_controller_cap`,
        arguments: [identityArg, action, receiving],
    });
    intentFn(tx, borrowedControllerCap);
    tx.moveCall({
        target: `${packageId}::controller_proposal::put_back`,
        arguments: [action, borrowedControllerCap],
    });
    return tx.build();
}
