// @iota/iota-interaction-ts/web/move_calls/asset/transfer.js
// Copyright 2020-2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0
import { Transaction } from "@iota/iota-sdk/transactions";
export function transfer(asset, assetType, recipient, packageId) {
    const tx = new Transaction();
    const assetArg = tx.objectRef(asset);
    const recipientArg = tx.pure.address(recipient);
    tx.moveCall({
        target: `${packageId}::asset::transfer`,
        typeArguments: [assetType],
        arguments: [assetArg, recipientArg],
    });
    return tx.build();
}
function makeTx(proposal, cap, asset, assetType, packageId, functionName) {
    const tx = new Transaction();
    const proposalArg = tx.sharedObjectRef(proposal);
    const capArg = tx.objectRef(cap);
    const assetArg = tx.objectRef(asset);
    tx.moveCall({
        target: `${packageId}::asset::${functionName}`,
        typeArguments: [assetType],
        arguments: [proposalArg, capArg, assetArg],
    });
    return tx.build();
}
export function acceptProposal(proposal, recipientCap, asset, assetType, packageId) {
    return makeTx(proposal, recipientCap, asset, assetType, packageId, "accept");
}
export function concludeOrCancel(proposal, senderCap, asset, assetType, packageId) {
    return makeTx(proposal, senderCap, asset, assetType, packageId, "conclude_or_cancel");
}
