// @iota/iota-interaction-ts/node/move_calls/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getClockRef = getClockRef;
exports.getControllerDelegation = getControllerDelegation;
exports.putBackDelegationToken = putBackDelegationToken;
exports.insertPlaceholders = insertPlaceholders;
const utils_1 = require("@iota/iota-sdk/utils");
const PLACEHOLDER_SENDER = "0x00000000000000090807060504030201";
const PLACEHOLDER_GAS_BUDGET = 9;
const PLACEHOLDER_GAS_PRICE = 8;
const PLACEHOLDER_GAS_PAYMENT = [];
function getClockRef(tx) {
    return tx.sharedObjectRef({ objectId: utils_1.IOTA_CLOCK_OBJECT_ID, initialSharedVersion: 1, mutable: false });
}
function getControllerDelegation(tx, controllerCap, packageId) {
    const [token, borrow] = tx.moveCall({
        target: `${packageId}::controller::borrow`,
        arguments: [controllerCap],
    });
    return [token, borrow];
}
function putBackDelegationToken(tx, controllerCap, delegationToken, borrow, packageId) {
    tx.moveCall({
        target: `${packageId}::controller::put_back`,
        arguments: [controllerCap, delegationToken, borrow],
    });
}
/**
 * Inserts placeholders related to sender and payment into transaction.
 *
 * This is required if wanting to call `tx.build`, as this will check if these values have been set.
 *
 * @param tx transaction to update
 */
function insertPlaceholders(tx) {
    tx.setGasPrice(PLACEHOLDER_GAS_PRICE);
    tx.setGasBudget(PLACEHOLDER_GAS_BUDGET);
    tx.setGasPayment([...PLACEHOLDER_GAS_PAYMENT]); // make sure, we're not sharing the array between tx
    tx.setSender(PLACEHOLDER_SENDER);
}
